File: Algebras/Rose/RoseJava/Element.java
//This file is part of SECONDO.

//Copyright (C) 2004, University in Hagen, Department of Computer Science, 
//Database Systems for New Applications.

//SECONDO is free software; you can redistribute it and/or modify
//it under the terms of the GNU General Public License as published by
//the Free Software Foundation; either version 2 of the License, or
//(at your option) any later version.

//SECONDO is distributed in the hope that it will be useful,
//but WITHOUT ANY WARRANTY; without even the implied warranty of
//MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//GNU General Public License for more details.

//You should have received a copy of the GNU General Public License
//along with SECONDO; if not, write to the Free Software
//Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import java.io.*;
//import java.util.*;

abstract public class Element implements Serializable{
  //Abstract class for all geometric objects, particularly Point, Segment,
  //Triangle, Polygon and all user-defined objects

  //It lists all methods which have to be implemented
  //by all classes which define primitives, it is especially
  //important for the implementation of new primitives
  //since it shows a complete list of essential 
  //methods for that new primitive.


  //members
  //public double area = 0;
  //public double perimeter = 0;
  //public Rect bbox = new Rect(); //bounding box

  //constructors
  //Element(LinkedList vertices);
  //creates a new object with given name and given vertices


  //methods
  abstract Element copy();
  //returns a deep copy of THIS.object as new object

 
  abstract boolean equal(Element inObject) throws WrongTypeException;
  //returns TRUE, if the given object and THIS.object are congruent
  
  abstract byte compX(Element inObject) throws WrongTypeException;
  //compares the x-coordinates of the leftmost points of the
  //given object and THIS.object and returns
  //-1, if THIS.object has a smaller x-coordinate
  //0, if the x-coordinates are equal
  //+1, if THIS.object has a greater x-coordinate

  abstract byte compY(Element inObject) throws WrongTypeException;
  //similar to compx but uses y-coordinates

    abstract byte compare(Element inObject) throws WrongTypeException;
    //uses both, compX and compY

  //void set(String name; LinkedList vertices);
  //sets THIS.name to name and the THIS.object's vertices to
  //this vertices

  //abstract void update();
  //forces a new computation of class members, e.g. area and perimeter

  //double compute_perimeter();
  //returns the perimeter of THIS.object

  //double compute_area();
  //returns the area of THIS.object

  //String isA();
  //returns class name of THIS.object as String

  //boolean isValid();
  //TRUE, if THIS.object is valid due to defined (user-)restriction

  //boolean buildOfTriangles();
  //TRUE, if THIS.object may be parted to a set of triangles

  //LinkedList getTriangles();
  //returns a LinkedList of triangles forming THIS.object, if
  //buildOfTriangles returns TRUE
  //all elements of LinkedList must be of type Triangle

  //abstract void computeBbox();
  //computes the bounding box and sets bbox

  abstract void print();
  //prints the elements data to text window

  abstract boolean intersects(Element inObject) throws WrongTypeException;
  //returns true if this.object an in intersect
  //false else

    abstract Rect rect();
    //returns the bounding box of this.object

    abstract Rational dist(Element inObject) throws WrongTypeException;
    //returns the distance between two Elements of the same type

}//end class Element

File: Algebras/OSM/Element.cpp
/*
----
This file is part of SECONDO.

Copyright (C) 2004, University in Hagen, Department of Computer Science,
Database Systems for New Applications.

SECONDO is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

SECONDO is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with SECONDO; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
----

//paragraph [1] Title: [{\Large \bf \begin {center}] [\end {center}}]
//[TOC] [\tableofcontents]
//[_] [\_]

[1] Implementation of the OSM Algebra

June-November, 2011. Thomas Uchdorf

[TOC]

1 Overview

This implementation file contains the implementation of the class
~Element~.

For more detailed information see Element.h.

2 Defines and Includes

*/
// [...]
#undef __TRACE__
//#define __TRACE__ cout <<  __FILE__ << "::" << __LINE__;
#define __TRACE__

// --- Including header-files
#include "Element.h"
#include <iostream>
#include <cassert>

// --- Constructors
// Default-Constructor
Element::Element ()
  : m_name (), m_attributeNames (), m_attributeValues (), m_value (),
    m_level (0)
{
   // empty
}
// Constructor
Element::Element (std::string name, 
                  std::vector<std::string> attributeNames,
                  std::vector<std::string> attributeValues,
                  std::string value,
                  int level)
  : m_name (name), m_attributeNames (attributeNames),
  m_attributeValues (attributeValues), m_value (value), m_level (level)
{
   // empty
}
// Destructor
Element::~Element ()
{
   // empty
}

// --- Methods
void Element::setName (const std::string &name)
{
    m_name = name;
}

void Element::setValue (const std::string &value)
{
    m_value = value;
}

void Element::setAttributeNames (const std::vector<std::string> &attributeNames)
{
    m_attributeNames = attributeNames;
}

void Element::setAttributeValues (const std::vector<std::string> &
    attributeValues)
{
    m_attributeValues = attributeValues;
}

void Element::setLevel (const int &level)
{
    m_level = level;
}

const std::string & Element::getName () const
{
    return m_name;
}

const std::string & Element::getValue () const
{
    return m_value;
}

const std::vector<std::string> & Element::getAttributeNames () const
{
    return m_attributeNames;
}

const std::vector<std::string> & Element::getAttributeValues () const
{
    return m_attributeValues;
}

const int & Element::getLevel () const
{
    return m_level;
}

void Element::print () const
{
    printElement (*this);
}

void printElement (const Element &element)
{
    // Printing the processed element
    std::cout << element.getName () << std::endl;
    std::cout << "=" << element.getValue () << std::endl;
    std::vector<std::string>::const_iterator itAttrNames;
    std::vector<std::string>::const_iterator itAttrValues;
    std::vector<std::string> const & attributeNames = 
        element.getAttributeNames ();
    std::vector<std::string> const & attributeValues = 
        element.getAttributeValues ();
    assert (attributeNames.size () == attributeValues.size ());
    for (itAttrNames = attributeNames.begin (),
         itAttrValues = attributeValues.begin ();
         itAttrNames != attributeNames.end ();
         ++itAttrNames,++itAttrValues)  {
        std::cout << "- " << (*itAttrNames) << " = " << (*itAttrValues);
        std::cout << std::endl;
    }
}
